### PR TITLE
docker: Override GOMODCACHE to always use a writable location

### LINKED ISSUE #6354: Investigate spurious package download errors when running in CI
I'm seeing a lot of Go module downloads being attempted when running lint/test targets in CI. I'm pretty sure this shouldn't be happening on the main Teleport codebase as all dependencies should be vendored. Investigate why this is happening and fix it.

This may be related to the `api` directory (as I think we intend that dependencies should be downloaded for that package specifically) and if so, we should modify CI to write the cache to a writable temp dir to remove the errors.

```
docker run --rm=true -v "$(pwd)/../":/go/src/github.com/gravitational/teleport -v /tmp:/tmp -w /go/src/github.com/gravitational/teleport -h buildbox -v /tmpfs/go/kubeconfig.ci:/mnt/kube/config -e KUBECONFIG=/mnt/kube/config -e TEST_KUBE=true -v /tmpfs/go/cache:/go/cache -e GOCACHE=/go/cache -v "$(pwd)/bcc:/usr/include/bcc" -u 1000:1000 -t quay.io/gravitational/teleport-buildbox:go1.16.2 \
	/bin/bash -c "make -C /go/src/github.com/gravitational/teleport FLAGS='-cover' integration"
make: Entering directory '/go/src/github.com/gravitational/teleport'
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/jonboulle: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/gopkg.in: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/niemeyer: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/google: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/gogo: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/sirupsen: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/stretchr: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/gravitational: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/net/@v/v0.0.0-20190311183353-d8887717615a.mod496193015.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/konsorten: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/golang.org/x/oauth2: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/gopkg.in: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/net/@v/v0.0.0-20200707034311-ab3426394381.mod277341737.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/google: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/gopkg.in: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/kr: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/golang.org/x/lint: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/tools/@v/v0.0.0-20190226205152-f727befe758c.mod168565194.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/envoyproxy: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/davecgh: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/kr: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/github.com/kisielk/errcheck/@v/v1.2.0.mod890951957.tmp: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/sys/@v/v0.0.0-20190422165155-953cdadca894.mod920627189.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/prometheus: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/client9: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/honnef.co: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/golang.org/x/exp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/cloud.google.com: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/sys/@v/v0.0.0-20200323222414-85ca7c5b95cd.mod568160156.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/!burnt!sushi: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/net/@v/v0.0.0-20190213061140-3a22650c66bd.mod926275561.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/pmezard: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/sync/@v/v0.0.0-20181108010431-42b317875d0f.mod980435746.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/envoyproxy: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/golang.org/x/lint: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/tools/@v/v0.0.0-20181030221726-6c7e314b6563.mod990689002.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/census-instrumentation: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/sync/@v/v0.0.0-20180314180146-1d60e4601c6f.mod526661577.tmp: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/sys/@v/v0.0.0-20180830151530-49385e6e1522.mod210707463.tmp: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/tools/@v/v0.0.0-20190114222345-bf090417da8b.mod394767996.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/net/@v/v0.0.0-20180826012351-8a410e7b638d.mod623118623.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/golang.org/x/lint: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/tools/@v/v0.0.0-20190524140312-2c0ae7006135.mod924879241.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/honnef.co: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/net/@v/v0.0.0-20180724234803-3673e40ba225.mod548653033.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: open /go/pkg/mod/cache/download/golang.org/x/tools/@v/v0.0.0-20190311212946-11955173bddd.mod183653891.tmp: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/golang: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/google: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/google.golang.org: permission denied
go: writing go.mod cache: mkdir /go/pkg/mod/cache/download/github.com/google: permission denied
go: downloading github.com/gravitational/trace v1.1.13
go: downloading github.com/gogo/protobuf v1.3.1
go: downloading golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
go: downloading github.com/golang/protobuf v1.4.2
go: downloading google.golang.org/grpc v1.27.0
go: downloading golang.org/x/net v0.0.0-20200707034311-ab3426394381
go: downloading gopkg.in/yaml.v2 v2.4.0
KUBECONFIG is: /mnt/kube/config, TEST_KUBE: true
go test -tags "pam  bpf" github.com/gravitational/teleport/e/integration github.com/gravitational/teleport/integration -cover
ok  	github.com/gravitational/teleport/e/integration	0.061s	coverage: [no statements]\
```

### PR DESCRIPTION
**Pull request by webvictim**
A combination of our refactoring to keep our API client inside a separate `api` subdirectory (not using a `vendor` directory) and switching to Go 1.16 has caused a lot of package downloads to be attempted inside the Docker container when running commands like `make -C build.assets release` and `make -C build.assets test`.

As `/go/pkg/mod` (default value for `GOMODCACHE`) is owned by `root` inside the buildbox container and not writable by a normal user, these package downloads fail when the Docker container is run as `$(uid):$(gid)` (as it is by default). These failures are benign as Go falls back to using a `.cache` directory inside the working directory if it can't write to `GOMODCACHE`, but they make debugging harder.

This PR changes Docker builds to use `/tmp/gomodcache` as the cache directory. As `/tmp` is mounted chmod 777 with the sticky bit on the host (and our `build.assets/Makefile` mounts it into the container) any user is able to write files there. This means we don't need to worry about adding a separate user or changing UID/GID inside the container, so we avoid all the associated annoyances that can be caused when the UID/GID inside a Docker container don't match those used outside.

The one change with this (a possible downside) is that `/tmp/gomodcache` may keep a persistent cache when Docker builds are run repeatedly. As Go validates the hash of downloaded modules when running this should not cause issues. Just in case, though, `make -C build.assets clean` has been updated to also run `go clean -modcache` on `/tmp/gomodcache`. This issue will not affect CI because CI containers get a brand new `/tmp` for each run.

Fixes #6354

### CODE REVIEW

File: build.assets/Dockerfile [hunk 1]
```diff
@@ -71,6 +71,7 @@ RUN mkdir -p /opt && cd /opt && curl https://storage.googleapis.com/golang/$RUNT
 
 ENV GOPATH="/go" \
     GOROOT="/opt/go" \
+    GOMODCACHE="/tmp/gomodcache" \
```
Comment (wadells) [inline build.assets/Dockerfile:74]: I'd put this as a `-e` flag where `/tmp` is mounted to the container, to keep like code/concerns grouped.

**Review by wadells — COMMENTED** [review]
Does the `lint-api` target needs its `GOMODCACHE` declaration cleaned up too?

File: build.assets/Dockerfile (hunk 1 above)
Comment (webvictim) [reply build.assets/Dockerfile]: Good idea. Addressed in 6d62d2d.

**Review by wadells — APPROVED** [review]

**Review by awly — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**